Jamulus.pro

```diff
@@ -109,6 +109,51 @@ win32 {
         INCLUDEPATH += "C:/Program Files (x86)/Jack/includes"
         LIBS += "C:/Program Files (x86)/Jack/lib/libjack64.lib"
     }
+} else:win64 {
+    DEFINES -= UNICODE # fixes issue with ASIO SDK (asiolist.cpp is not unicode compatible)
+    DEFINES += NOMINMAX # solves a compiler error in qdatetime.h (Qt5)
+    HEADERS += windows/sound.h
+    SOURCES += windows/sound.cpp \
+        windows/ASIOSDK2/common/asio.cpp \
+        windows/ASIOSDK2/host/asiodrivers.cpp \
+        windows/ASIOSDK2/host/pc/asiolist.cpp
+    RC_FILE = windows/mainicon.rc
+    INCLUDEPATH += windows/ASIOSDK2/common \
+        windows/ASIOSDK2/host \
+        windows/ASIOSDK2/host/pc
+    mingw* {
+        LIBS += -lole32 \
+            -luser32 \
+            -ladvapi32 \
+            -lwinmm \
+            -lws2_32
+    } else {
+        QMAKE_LFLAGS += /DYNAMICBASE:NO # fixes crash with libjack64.dll, see https://github.com/jamulussoftware/jamulus/issues/93
+        LIBS += ole32.lib \
+            user32.lib \
+            advapi32.lib \
+            winmm.lib \
+            ws2_32.lib
+    }
+
+    # replace ASIO with jack if requested
+    contains(CONFIG, "jackonwindows") {
+        message(Using Jack instead of ASIO.)
+
+        !exists("C:/Program Files/JACK2/includes/jack/jack.h") {
```

Comment:
If possible, it may make sense to define some JACK_BASE_PATH variable and derive all other paths from that?

Reply:
I think it would work like this:
```
        contains(QT_ARCH, "x86") {
            jack_dir = "C:/Program Files (x86)/Jack"
        } else {
            jack_dir = "C:/Program Files/Jack"
        }

        !exists($${jack_dir}/includes/jack/jack.h) {
            message(Warning: jack.h was not found at the usual place, maybe jack is not installed)
        }
```

```
        INCLUDEPATH += $${jack_dir}/includes"
        LIBS += $${jack_dir}/lib/libjack64.lib
```

Reply:
do you think it is possible to use Windows environment variable instead of the hard coded one ?
Something like : `%ProgramFiles%` or the `ProgramFilesDir` key used in registry ?

Reply:
To access `%ProgramFiles(x86)%` we would need [`getenv`](https://doc.qt.io/qt-5/qmake-function-reference.html#getenv-variablename), which requires Qt 5.12 (might be okay, I think there's already a proposal to bump min Qt version for Mac).

```
        contains(QT_ARCH, "x86") {
            jack_dir = $$getenv("ProgramFiles(x86)")/Jack
        } else {
            jack_dir = $$getenv("ProgramFiles")/Jack
        }
```

Reply:
Autobuild script for Windows is using Qt 5.15.2 already as well, so it should not be a problem to use getenv.

Reply:
> which requires Qt 5.12 (might be okay, I think there's already a proposal to bump min Qt version for Mac).

I think we only agreed to bump the Qt version for the official builds. We did not drop support for building with older versions of Qt, as far as I know.

Reply:
@jujudusud the line:  !exists("C:/Program Files/JACK2/includes/jack/jack.h")
is incorrect referring to /includes/ here...it should be /include/  

Reason I did not notice is because the "message" function is throwing an error.

 message(Warning: jack.h was not found at the usual place, maybe jack is not installed)

This is generating an error due to the , (comma) in the text. The error says that message requires one argument.
Either replace the , by another character or use "" around the text in the message.
Also as it is just a "message" it does not even stop the build process.

This "message" issue also appears in the "macx" section. Perhaps this can be adapted as well? Assuming the title of the PR should change in that case.

Reply:
What about 'read_registry(tree, key[, flag])' instead of getenv ?
'read_registry(HKLM, ProgramFilesDir[, 64])'
Should return the good path for 64bits.

Reply:
My view is...why make it so complex? As JACK might even have been installed somewhere else....I would say it is a slight overkill to include all this. Not that is isn't nice to have it...

Reply:
> My view is...why make it so complex? As JACK might even have been installed somewhere else..

I agree. The compromise I'd aim for would be: Make sure the default works and provide a way to override for builds against different paths (`qmake JACK_INSTALL_DIR=/path/to/jack` -- maybe there are existing conventions what the variable should be named).

Reply:
> qmake JACK_INSTALL_DIR=/path/to/jack

That would also be beneficial in case we want to setup an automated build. Not sure how such script would install jack...but with the variable it could become an option to consider.

Reply:
this is the use of read registry
`    # replace ASIO with jack if requested
    contains(CONFIG, "jackonwindows") {
        message(Using Jack instead of ASIO.)

        contains(QT_ARCH, "x86") {
        	ProgramFilesDir = read_registry(HKLM, ProgramFilesDir[, 32])
		    !exists("$${ProgramFilesDir}/JACK2/include/jack/jack.h") {
                message(Warning: "jack.h was not found at the usual place, maybe JACK2 is not installed")
            }

            HEADERS -= windows/sound.h
            SOURCES -= windows/sound.cpp
            HEADERS += linux/sound.h
            SOURCES += linux/sound.cpp
            DEFINES += WITH_JACK
            DEFINES += JACK_REPLACES_ASIO
            DEFINES += _STDINT_H # supposed to solve compilation error in systemdeps.h
            INCLUDEPATH += "$${ProgramFilesDir}/JACK2/include"
            LIBS += "$${ProgramFilesDir}/JACK2/lib/libjack.lib"
        }

        contains(QT_ARCH, "x86_64") {
		    ProgramFilesDir = read_registry(HKLM, ProgramFilesDir[, 64])
		    !exists("$${ProgramFilesDir}/JACK2/include/jack/jack.h") {
                message(Warning: "jack.h was not found at the usual place, maybe JACK2 64-bit is not installed")
            }

            HEADERS -= windows/sound.h
            SOURCES -= windows/sound.cpp
            HEADERS += linux/sound.h
            SOURCES += linux/sound.cpp
            DEFINES += WITH_JACK
            DEFINES += JACK_REPLACES_ASIO
            DEFINES += _STDINT_H # supposed to solve compilation error in systemdeps.h
            INCLUDEPATH += "$${ProgramFilesDir}/JACK2/include"
            LIBS += "$${ProgramFilesDir}/JACK2/lib/libjack64.lib"
        }`

Reply:
ok ! think  it will be ok with 
c0a5110